### DISCUSSION TITLE
`ui.notify` from a background process?

### DISCUSSION THREAD

**Discussion by adarshpunj** [zauberzeug/nicegui#2893] [Q&A] [answered]
### Update: I was able to use `ui.refreshable` as a hack - however, I'm not sure if this is a good solution.
-------------------
### Question

Hi, 

I'm trying to run a function with ui.notify like this:

```python
def executor(arg):
    response = some_task(arg) # i can't modify this `some_task` callable; can be both io or cpu bound
    if not response:
        ui.notify("Failed successfully")

ui.button("Click me", on_click=lambda: run.io_bound(executor, arg=arg)

```
This throws:
```shell
RuntimeError: The current slot cannot be determined because the slot stack for this task is empty.
This may happen if you try to create UI from a background task.
To fix this, enter the target slot explicitly using `with container_element:`.
```
I can't do `on_click=lambda: executor(arg=arg)` as this will block the UI for the time `some_task` runs.

Normally, I'd use webhooks in cases like this (especially when I'm able to modify the `some_task` method)

What are some Pythonic solutions to this problem?

**Comment by falkoschindler** [comment]
Hi @adarshpunj,

The problem is that `ui.notify` doesn't know which client to notify. There could be multiple pages with multiple connected browsers.

But you can try a similar approach like we do in [RoSys](https://rosys.io/):
```py
def notify(message: str) -> None:
    for client in Client.instances.values():
        if not client.has_socket_connection:
            continue
        with client:
            ui.notify(message)
```
https://github.com/zauberzeug/rosys/blob/c87eb713b4bc7324e78524ae2420a365b7c1b5cb/rosys/rosys.py#L61-L80

This sends the notification to _all_ connected clients.

**Reply by adarshpunj** [reply]
Thanks for sharing. 

What do you think of this solution using `ui.refreshable`

```python
@ui.refreshable
def notify(msg: str | None) -> None:
    if msg:
        ui.notify(msg)

def executor(arg):
    response = some_task(arg)
    if not response:
        notify.refresh("Failed successfully")

notify(None)

```

**Reply by falkoschindler** [reply, marked as answer]
Interesting solution! It might be less intuitive to read, but should work as well. This way you're sending the notification only to the client who called `notify(None)` in the first place. 👍🏻
